tests/README.md

```diff
@@ -0,0 +1,1510 @@
+1. [Continues integration](#continues-integration)
+   1. [Test suites](#test-suites)
+   1. [Benefits](#benefits)
+   1. [Current problems](#current-problems)
+      1. [External PHPUnit version](#external-phpunit-version)
+      1. [Risky prebuild step](#risky-prebuild-step)
+1. [External Documentation](#external-documentation)
+1. [Old Documentation](#old-documentation)
+1. [FAQ](#faq)
+   1. [What needs to be tested ?](#what-needs-to-be-tested-)
+   1. [How is a local test environment set up ?](#how-is-a-local-test-environment-set-up-)
+   1. [How do I start the ILIAS unit tests in PHPStorm ?](#how-do-i-start-the-ilias-unit-tests-in-phpstorm-)
+   1. [How do I start the ILIAS unit tests in CLI ?](#how-do-i-start-the-ilias-unit-tests-in-cli-)
+   1. [Where do I put my unit tests ?](#where-do-i-put-my-unit-tests-)
+   1. [What is the ILIAS CI-Server and how can I benefit from it in terms of unit tests ?](#what-is-the-ilias-ci-server-and-how-can-i-benefit-from-it-in-terms-of-unit-tests-)
+   1. [What do I need to consider concerning unit tests before pushing code to the ILIAS repo ?](#what-do-i-need-to-consider-concerning-unit-tests-before-pushing-code-to-the-ilias-repo-)
+1. [Sources](#sources)
+
+<!-- /MarkdownTOC -->
+
+
+<a name="definition"></a>
+## Definition
+Unit testing is the process of testing the smallest testable part of an application, called units.
+Units are usually one behaviour of a class or one logical concept of a method.
+
+Unit tests fulfil the following criteria:
+- Able to be **fully automated**
+- **Isolated** (Dependencies of the class under test are mocked)
+- Runnable in **any order**, if the test is part of many other tests
+- **In memory** (no db or filesystem access)
+- **Consistent** results (Should always return the same, no random numbers)
+- Runs **fast** (Test should at most take 1 seconds to finish)
+- Tests a **single logical unit**
+- **Readable**
+- **Maintainable**
+- **Trustworthy** (The result of the test must be correct)
+
+<a name="role-in-ilias"></a>
+### Role in ILIAS
+
+All unit tests in ILIAS are fully automated with the goal to verify that all units behave as intended.
+Furthermore, unit tests enable the community to tackle bugs much faster and in an earlier state of development.
+Due to a faster handling of side effects and other bugs, more time is left to actually refactor and improve the current code base.
+
+<a name="tools"></a>
+## Tools
+<a name="php-unit"></a>
+### PHP Unit
+PHP Unit is a collection of tools (PHP classes and executables) which makes not only testing easy, but also helps to gain
+insight into the test results and how much of the code base remains untested.
+
+<a name="mockery"></a>
+### Mockery
+Mockery is a lightweight and flexible mocking object framework which is used for unit testing with PHP Unit and other unit testing frameworks.
+It is designed as a drop in replacement for the PHP Unit mock functionality, but can also work alongside with the PHP Unit mock objects.
```

Comment:
We should make a final official decision about mockery in the JF.

Reply:
@alex40724 was the ILIAS mocking solution voting not official ?
<https://www.ilias.de/docu/goto.php?target=cat_582>

Reply:
See our guideline/rule at
https://github.com/ILIAS-TechnicalBoard/ILIAS/tree/trunk/libs/composer

Usually these votings are made to get a picture of general preferences of groups of developers or users. They are not an official part of our development processes. I am not aware that the Jour Fixe has put the final decision into the hand of a voting in this case.

I think adding this particular dependency is a very important one, since lot of code/tests will depend on it in the future.

Reply:
@alex40724 Thx for the clarification.
I also spoke with @chfsx about the final decision, he will add it to the agenda.